openml/_api_calls.py

```diff
@@ -198,17 +188,42 @@ def __send_request(
                     response = session.post(url, data=data, files=files)
                 else:
                     raise NotImplementedError()
-                break
+                pass
             except (requests.exceptions.ConnectionError, requests.exceptions.SSLError,) as e:
                 if i == n_retries:
                     raise e
                 else:
                     time.sleep(0.1 * i)
+                    continue
+
+            try:
+                __check_response(response=response, url=url, file_elements=files)
+            except OpenMLServerException as e:
+                # Do retries in case of a database connection error
+                if e.code == 107:
+                    if i == n_retries:
+                        raise e
+                    else:
+                        time.sleep(i * 0.3)
+                        continue
+                else:
+                    raise e
```

Comment:
Maybe simplify the flow by calling `check_response` where you currently have the `pass` statement. You can then catch the `OpenMLServerException` in the same `except` clause and use shared logic on whether or not an error should be raised or a new connection attempt should be made.

Reply:
You mean like this?
```
            try:
                if request_method == "get":
                    response = session.get(url, params=data)
                elif request_method == "delete":
                    response = session.delete(url, params=data)
                elif request_method == "post":
                    response = session.post(url, data=data, files=files)
                else:
                    raise NotImplementedError()
                __check_response(response=response, url=url, file_elements=files)
                break
            except (
                requests.exceptions.ConnectionError, 
                requests.exceptions.SSLError,
                OpenMLServerException,
            ) as e:
                if isinstance(e, OpenMLServerException) and e.code != 107:
                    # 107 is a database connection error - only then do retries
                    raise e
                if i == n_retries:
                    raise e
                else:
                    time.sleep(0.1 * i)
                    continue
```
Or a bit more verbose like that:
```
            try:
                if request_method == "get":
                    response = session.get(url, params=data)
                elif request_method == "delete":
                    response = session.delete(url, params=data)
                elif request_method == "post":
                    response = session.post(url, data=data, files=files)
                else:
                    raise NotImplementedError()
                __check_response(response=response, url=url, file_elements=files)
                break
            except (
                requests.exceptions.ConnectionError,
                requests.exceptions.SSLError,
            ) as e:
                if i == n_retries:
                    raise e
                else:
                    time.sleep(0.1 * i)
                    continue
            except OpenMLServerException as e:
                # 107 is a database connection error - only then do retries
                if e.code != 107:
                    raise
                if i == n_retries:
                    raise e
                else:
                    time.sleep(0.3 * i)
                    continue
```
In any case, merging the two try/excepts is an improvement to the code.

Reply:
Fine with either, but I think I prefer the first one (with a `wait_time` variable if you want to preserve the difference in sleep time).